File: Sean的leetcode/Leetcode/Remove Duplicate Letters.cpp
/* Greedy
 * Time Complexity: O(n)
 * cnt: number of each character
 * visit: character is appear in the res string
 */
class Solution {
public:
    string removeDuplicateLetters(string s) {
        string res;
        vector<int> cnt(256, 0);
        vector<bool> visit(256, false);
        for (auto c : s) {
            cnt[c]++;
        }
        for (auto c : s) {
            cnt[c]--;
            if (visit[c]) continue;
            while (!res.empty() && c < res.back() && cnt[res.back()]) {
                visit[res.back()] = false;
                res.pop_back();
            }
            visit[c] = true;
            res += c;
        }
        return res;
    }
};

File: MyLeetCode/python/Remove Duplicate Letters.py
###
# Greedy Stack
# Time Complexity: O(n)
# Space Complexity: O(n)
###
class Solution(object):
    def removeDuplicateLetters(self, s):
        """
        :type s: str
        :rtype: str
        """
        if not s:
            return s
        from collections import Counter
        cnt = Counter(s)
        visit = set()
        re = []
        for c in s:
            cnt[c] -= 1
            if c in visit:
                continue
            while re and re[-1] > c and cnt[re[-1]]:
                visit.remove(re[-1])
                re.pop()
            visit.add(c)
            re.append(c)     
        return ''.join(re)
        
                                        
                            
                    